File: Problems/Algorithms/560. Subarray Sum Equals K/SubarrayK.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        int ans = 0;
        int prefix = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        
        map.put(0, 1);
        for (int i = 0; i < nums.length; i++) {
            prefix += nums[i];
            if (map.containsKey(prefix-k)) {
                ans += map.get(prefix-k);
            }
            map.put(prefix, map.getOrDefault(prefix, 0) + 1);
        }
        
        return ans;
    }
}

File: Problems/Algorithms/560. Subarray Sum Equals K/SubarrayK.kt
class Solution {
    fun subarraySum(nums: IntArray, k: Int): Int {
        var ans = 0
        var prefix = 0
        val map = mutableMapOf<Int, Int>()
        
        map.put(0, 1)
        for (i in 0..nums.size-1) {
            prefix += nums[i]
            if (map.containsKey(prefix-k)) {
                ans += map.get(prefix-k)!!
            }
            map.put(prefix, map.getOrDefault(prefix, 0) + 1)
        }
        
        return ans
    }
}

